lib/versioner/versioner_test.go

```diff
@@ -0,0 +1,103 @@
+// Copyright (C) 2020 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package versioner
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/syncthing/syncthing/lib/config"
+	"github.com/syncthing/syncthing/lib/fs"
+)
+
+func TestVersionerCleanOut(t *testing.T) {
+	simpleCfg := config.FolderConfiguration{
+		FilesystemType: fs.FilesystemTypeBasic,
+		Path:           "testdata",
+		Versioning: config.VersioningConfiguration{
+			Params: map[string]string{
+				"cleanoutDays": "7",
+			},
+		},
+	}
+
+	trashCfg := config.FolderConfiguration{
+		FilesystemType: fs.FilesystemTypeBasic,
+		Path:           "testdata",
+		Versioning: config.VersioningConfiguration{
+			Params: map[string]string{
+				"cleanoutDays": "7",
+			},
+		},
+	}
+	testCasesVersioner := []Versioner{
+		newSimple(simpleCfg),
+		newTrashcan(trashCfg).(*trashcan),
+	}
+
+	var testcases = []struct {
+		file         string
+		shouldRemove bool
+	}{
+		{"testdata/.stversions/file1", false},
+		{"testdata/.stversions/file2", true},
+		{"testdata/.stversions/keep1/file1", false},
+		{"testdata/.stversions/keep1/file2", false},
+		{"testdata/.stversions/keep2/file1", false},
+		{"testdata/.stversions/keep2/file2", true},
+		{"testdata/.stversions/keep3/keepsubdir/file1", false},
+		{"testdata/.stversions/remove/file1", true},
+		{"testdata/.stversions/remove/file2", true},
+		{"testdata/.stversions/remove/removesubdir/file1", true},
+	}
+
+	for index, versioner := range testCasesVersioner {
+		t.Run(fmt.Sprintf("versioner trashcan clean up %d in %d", index, len(testCasesVersioner)), func(t *testing.T) {
```

Comment:
I'd rather store the versioners in a map using names as key and use those for logging here - more informative than e.g. "0 in 2".

Reply:
I cannot pass the filename to the logging in t.Run(string, func(){}) since the looping of the file occurs inside the body of the t.Run function. Can you point me to an example of logging in syncing?

Reply:
What I mean is changing the current slice to a map like this

```
	testCasesVersioner := map[string]Versioner{
		"simple": newSimple(simpleCfg),
		"trashcan"; newTrashcan(trashCfg),
```

and then

```
	for versionerType, versioner := range testCasesVersioner {
		t.Run(fmt.Sprintf("%v versioner clean up", versionerType, len(testCasesVersioner)), func(t *testing.T) {
```